lib/entities/quote/DutchQuoteFactory.ts

```diff
@@ -0,0 +1,158 @@
+import { TradeType } from '@uniswap/sdk-core';
+import { log } from '@uniswap/smart-order-router';
+import { BigNumber } from 'ethers';
+import { v4 as uuidv4 } from 'uuid';
+import { ChainConfigManager } from '../../config/ChainConfigManager';
+import { NATIVE_ADDRESS, QuoteType } from '../../constants';
+import { Portion } from '../../fetchers/PortionFetcher';
+import { generateRandomNonce } from '../../util/nonce';
+import { currentTimestampInMs } from '../../util/time';
+import { DutchQuoteRequest, DutchV1Request, DutchV2Request } from '../request';
+import { ClassicQuote } from './ClassicQuote';
+import { DutchQuote, DutchQuoteConstructorArgs, DutchQuoteJSON, ParameterizationOptions } from './DutchQuote';
+import { DutchV1Quote } from './DutchV1Quote';
+import { DutchV2Quote } from './DutchV2Quote';
+
+export class DutchQuoteFactory {
+  // build a dutch quote from an RFQ response
+  public static fromResponseBody(
+    request: DutchQuoteRequest,
+    body: DutchQuoteJSON,
+    nonce?: string,
+    portion?: Portion
+  ): DutchQuote<DutchQuoteRequest> {
+    // if it's exact out, we will explicitly define the amount out start to be the swapper's requested amount
+    const amountOutStart =
+      request.info.type === TradeType.EXACT_OUTPUT ? request.info.amount : BigNumber.from(body.amountOut);
+    const { amountIn: amountInEnd, amountOut: amountOutEnd } = DutchQuote.applySlippage(
+      { amountIn: BigNumber.from(body.amountIn), amountOut: amountOutStart },
+      request
+    );
+    const args: DutchQuoteConstructorArgs = {
+      createdAtMs: currentTimestampInMs(),
+      request,
+      chainId: body.chainId,
+      requestId: body.requestId,
+      quoteId: body.quoteId,
+      tokenIn: body.tokenIn,
+      tokenOut: body.tokenOut,
+      amountInStart: BigNumber.from(body.amountIn),
+      amountInEnd,
+      amountOutStart,
+      amountOutEnd,
+      swapper: body.swapper,
+      quoteType: QuoteType.RFQ,
+      filler: body.filler,
+      nonce,
+      portion,
+    };
+    if (request instanceof DutchV1Request) {
```

Comment:
ah I'm usually not a huge fan on instanceof checks, and this selection code is duplicated in other functions in this class - can we do something like `request.routingType` to differentiate these?

Reply:
Yeah, I really wanted to include these alt constructors in `DutchQuote` but I couldn't think of a good way to do it.